lib/spack/llnl/util/filesystem.py

```diff
@@ -1734,83 +1739,79 @@ def find(root, files, recursive=True, max_depth: Optional[int] = None):
-    try:
-        stat_root = os.stat(root)
-    except OSError as e:
-        if e.errno == errno.ENOENT:
-            return []
-        else:
-            raise
-    if not stat.S_ISDIR(stat_root.st_mode):
-        raise ValueError(f"{root} is not a directory")
 
-    if max_depth is None:
-        max_depth = sys.maxsize
-
-    if isinstance(globs, str):
-        globs = [globs]
-    # Apply normcase to regular expressions and to the filenames:
-    # this respects case-sensitivity semantics of different OSes
-    # (e.g. file search is typically case-insensitive on Windows)
-    regexes = [re.compile(fnmatch.translate(os.path.normcase(x))) for x in globs]
-
-    # Note later calls to os.scandir etc. return abspaths if the
-    # input is absolute, see https://docs.python.org/3/library/os.html#os.DirEntry.path
-    root = os.path.abspath(root)
+def find_max_depth(
+    roots: Union[str, List[str]], globs: Union[str, List[str]], max_depth: Optional[int] = None
+) -> List[str]:
+    """Given a set of non-recursive glob file patterns, finds all files matching those patterns up
+    to a maximum specified depth.
 
-    found_files = collections.defaultdict(list)
+    If a directory has a name which matches an input pattern, it will not be included in the
+    results.
 
-    def _dir_id(stat_info):
-        # Note: on windows, st_ino is the file index and st_dev
-        # is the volume serial number. See
-        # https://github.com/python/cpython/blob/3.9/Python/fileutils.c
-        return (stat_info.st_ino, stat_info.st_dev)
+    If ``max_depth`` is specified, does not search below that depth.
 
-    def _log_file_access_issue(e):
-        errno_name = errno.errorcode.get(e.errno, "UNKNOWN")
-        tty.debug(f"find must skip {dir_entry.path}: {errno_name} {str(e)}")
+    If ``globs`` is a list, files matching earlier entries are placed in the return value before
+    files matching later entries.
+    """
 
-    visited_dirs = set([_dir_id(stat_root)])
+    if not isinstance(globs, list):
+        globs = [globs]
+    # Apply normcase to file patterns and filenames to respect case insensitive filesystems.
+    regexes = [re.compile(fnmatch.translate(os.path.normcase(x))) for x in globs]
+    if not isinstance(roots, list):
+        roots = [roots]
+    # Ensure returned paths are always absolute.
+    roots = [os.path.abspath(r) for r in roots]
+    if max_depth is None:
+        max_depth = sys.maxsize
+    # Maps file pattern to list of matching paths, preserving order of the patterns.
+    found_files: Dict[re.Pattern, List[str]] = {pattern: [] for pattern in regexes}
+
+    # Each queue item stores the depth and path. This achieves a consistent traversal order by
+    # iterating through each directory in alphabetical order. This also traverses in BFS order to
+    # ensure finding the shortest path from any root to any file (in case of multiple shortest
+    # paths the one returned will be deterministic given the prior point).
```

Comment:
True in the sense of same input same output. [x, y] != [y, x]

Reply:
If we can make this more straightforwardly true, that would be ideal: can you update the comment, or sort the roots?

Reply:
Right now directories are visited in the order the user passed them. That makes a whole lot of sense to me. For example find([x/bin, x/sbin]) would first check x/bin. Would we ever merge find and find_first this is certainly desirable behavior.

Reply:
> Right now directories are visited in the order the user passed them. That makes a whole lot of sense to me.

That's fine by me, but it should be documented. 

> For example `find([./bin, ./sbin])` would first check ./bin

Yes, although in general that doesn't extend for all `find([x/, y/])`: it does not search all of `x/` before all of `y/` (that's true for bin/sbin though because they don't have subdirs); I think it's worth being careful about describing that in the docstring. It might help to separately describe single and multi-root semantics.

Reply:
Alternatively we could not document the exact order so we don't have to stick to it.